02-supp-call-stack.Rmd

```diff
@@ -0,0 +1,113 @@
+What it *doesn't* hold is the variables that were in the various stack frames.
+If we try to get the value of `temp` after our functions have finished running, R tells us that there's no such thing:
+
+```{r, error = TRUE}
+temp
+```
+
+> ## Tip {.callout}
+>
+> The explanation of the stack frame above was very general and the basic
+> concept will help you understand most languages you try to program with.
+> However, R has some unique aspects that can be exploited when performing
+> more complicated operations. We will not be writing anything that requires
+> knowledge of these more advanced concepts. In the future when you are
+> comfortable writing functions in R, you can learn more by reading the
+> [R Language Manual][man] or this [chapter][] from
+> [Advanced R Programming][adv-r] by Hadley Wickham. For context, R uses the
+> terminology "environments" instead of frames.
+
+[man]: http://cran.r-project.org/doc/manuals/r-release/R-lang.html#Environment-objects
+[chapter]: http://adv-r.had.co.nz/Environments.html
+[adv-r]: http://adv-r.had.co.nz/
+
+Why go to all this trouble? Well, here's a function called `span` that calculates the difference between the minimum and maximum values in an array:
+
+```{r}
+span <- function(a) {
+  diff <- max(a) - min(a)
+  return(diff)
+}
+
+dat <- read.csv(file = "data/inflammation-01.csv", header = FALSE)
+# span of inflammation data
+span(dat)
+```
+
+Notice `span` assigns a value to variable called `diff`. We might very well use a variable with the same name (`diff`) to hold the inflammation data:
+
+```{r}
+diff <- read.csv(file = "data/inflammation-01.csv", header = FALSE)
+# span of inflammation data
+span(diff)
+```
+
+We don't expect the variable `diff` to have the value `r span(diff)` after this function call, so the name `diff` cannot refer to the same variable defined inside `span` as it does in as it does in the main body of our program (which R refers to as the global environment).
+And yes, we could probably choose a different name than `diff` for our variable in this case, but we don't want to have to read every line of code of the R functions we call to see what variable names they use, just in case they change the values of our variables.
+
+The big idea here is [encapsulation](reference.html#encapsulation), and it's the key to writing correct, comprehensible programs.
+A function's job is to turn several operations into one so that we can think about a single function call instead of a dozen or a hundred statements each time we want to do something.
+That only works if functions don't interfere with each other; if they do, we have to pay attention to the details once again, which quickly overloads our short-term memory.
+
+> ## Challenge - Following the call stack {.challenge}
+>
+>  + We previously wrote functions called `fence` and `outside`.
+>    Draw a diagram showing how the call stack changes when we run the
+>    following:
+> ```{r, results="hide"}
```

Comment:
from the `knitr` documentation: `hide` hides results; this option only applies to normal R output (not warnings, messages or errors)

if we actually execute the below block of code in this document, we have an error: 
`## Error in eval(expr, envir, enclos): could not find function "outside"`

maybe use `eval=FALSE` instead?